Refactor session to avoid duplicate calls to apis

cesnietor opened the pull request:
### Description:
Refactors the way we were calling `/session` api since previous implementaiton was making multiple requests (up to six) which could mean high latency if that requests take time to resolve. Now we call it only once and only at initial load. 

### Changes:
- [x] created Session thunk to control Session API state 
- [x] Create new state in consoleSlice to store the session state (loading, done, etc)
- [x] Updated Loging.tsx to default to `/browser` to avoid redirecting first to `/` and then to `/browser` when logging in.  

Before:
<img width="537" alt="Screenshot 2023-06-12 at 10 50 46 PM" src="https://github.com/minio/console/assets/11819101/d793ea39-9a3e-4e03-bf69-2d8f6f4398fb">

After:
<img width="378" alt="Screenshot 2023-06-12 at 11 05 27 PM" src="https://github.com/minio/console/assets/11819101/262a5e91-1b79-49a2-86e3-e46c37827436">

If navigating between tabs it won't call session api.
<img width="318" alt="Screenshot 2023-06-13 at 1 59 49 PM" src="https://github.com/minio/console/assets/11819101/daecdead-6584-4fdb-84ba-a22ed3ba7ce7">


### Test Steps:
Scenarios:
- With a minio running on port 9000, make assets and then make the binary `make assets && make` then run it on port 9090 like:
```
CONSOLE_ACCESS_KEY=minioconsole
CONSOLE_SECRET_KEY=minioconsole
CONSOLE_MINIO_SERVER=http://localhost:9000
CONSOLE_DEV_MODE=on ./console server
```
- Can login with valid user and the amount of calls to `/session` api are not more than one.
- Navigate between tabs (we should call `/session` only ONCE at initial load of whole page)
- Anonymous access works without issues (create bucket, set anonymous access to it  [path `/`] and try to access to it without loging in). Everything should work just fine.

dvaldivia commented on portal-ui/src/ProtectedRoutes.tsx:
```diff
@@ -53,56 +52,16 @@ const ProtectedRoute = ({ Component }: ProtectedRouteProps) => {
+  // update location path and store it
+  useEffect(() => {
+    dispatch(fetchSession(pathname));
+  }, [dispatch, pathname]);
```
This effect will cause us to check the session every time the path changes, if we already checked the session we shouldn't need to check it again

cesnietor replied:
I thought about that but if we are navigating through the app and you are logged in, what if someone wants to restrict your access while logged in, so calling it on every pathname would allow that. That's how it was before so assumed that was expected and kind of made sense.

cesnietor replied:
We discussed this and indeed is a nice "feature" that we have where we can hide the features they have access if we for some reason want to. But this is something we will sacrifice since calling /session api everytime the path changes is expensive.  
I've updated the PR to reflect this. It will now only call session at initial load. Anonymous access still works.

dvaldivia commented on portal-ui/src/screens/Console/consoleSlice.ts:
```diff
@@ -15,40 +15,56 @@
-import { SessionResponse } from "api/consoleApi";
+import { fetchSession } from "../../screens/LoginPage/sessionThunk";
+
+export enum SessionCallStates {
```
can we put this in `consoleSlice.types.ts` to keep the slice simple?

cesnietor replied:
we have types below but yes, I can bring them to a separate file, usually is better to keep related code closer to the code that uses it but we can have a types file.

cesnietor replied:
Actually I'd like to keep it there, is still simple don't you think?

dvaldivia replied:
I think it's better to keep it in a separate file like all the types we have everywhere else, slices tend to get large :P

cesnietor replied:
done.

dvaldivia commented on portal-ui/src/screens/LoginPage/sessionThunk.ts:
```diff
@@ -0,0 +1,95 @@
+            .finally(() => {
+              // TODO: we probably need a thunk for this api since setting the state here is hacky,
+              // we can use a state to let the ProtectedRoutes know when to render the elements
+              dispatch(setSessionLoadingState(SessionCallStates.Done));
```
do you really need to dispatch this? wouldn't the slice take care of this?

dvaldivia commented:
Seeing a session call every time I navigate @cesnietor 
<img width="513" alt="Screenshot 2023-06-13 at 9 21 21 AM" src="https://github.com/minio/console/assets/18384552/468c5069-db3d-40e8-9190-76363df900ad">

cesnietor commented:
@dvaldivia yes it was like that already before.

reivaj05 commented on portal-ui/src/screens/LoginPage/Login.tsx:
```diff
@@ -36,7 +36,8 @@ export interface LoginStrategyPayload {
 
 export const getTargetPath = () => {
-  let targetPath = "/";
+  // default to browser view to avoid redirecting first to "/" and then to "/browser" on login
```
Comment might not be necessary

cesnietor commented on portal-ui/src/ProtectedRoutes.tsx:
```diff
@@ -41,8 +34,14 @@ interface ProtectedRouteProps {
   const userLoggedIn = useSelector((state: AppState) => state.system.loggedIn);
+  // sessionLoading is used to control rendering this page, if we only use console.sessionLoadingState,
+  // since the state chages it triggers the reload of this component multiple times, causing duplicate queries.
+  const [sessionLoading, setSessionLoading] = useState<boolean>(true);
```
agreed, done.

cesnietor commented:
@prakashsvmx @reivaj05 @dvaldivia @jinapurapu  I've updated the PR addressing the comments and also doing the following -> We will no longer call `/session` api on every change path since it's a bit expensive (even if the api resolves quickly). More comments in the thread above with Daniel

cesnietor commented:
rebased.
